### DISCUSSION TITLE
Possibilty of adding multiple texture inputs for transitions.

### DISCUSSION THREAD

**Discussion by axiomgraph** [friction2d/friction#85] [Ideas]
I think adding multiple texture inputs will be useful for creating transitions. or it will help creating different kind of effects based on some image. so I left something for future.
![Screenshot from 2023-09-19 19-37-47](https://github.com/friction2d/friction/assets/68183694/97bb2c05-afb6-4d7f-93a2-49de8dfa320f)

**Comment by AlexKiryanov** [comment]
Hi all! By the way, I once proposed this idea to the developer Enve (Maurycy Liebner).
And yes, this is not a bad idea, because some shader effects require processing of two textures. You'll probably say that you can write the same texture as code. Yes it is. But this is not always convenient and can sometimes be a limitation.

**Reply by axiomgraph** [reply]
Yes, sometimes not having multiple textures is a limitation. Another problem is the OpenGL version the program is using. You have to create two codes. One uses the texture function, and one uses the texture2D function. or the problem is happening only for me on Windows 10; the texture() function works. On Arch Linux, the texture2D() function works.

**Comment by rodlie** [comment]
I understand the need, but the current implementation does not support this. Shader effects are isolated to the "box" (rectangle etc) in use.

But this is something to think about for the future.

I'm currently planning some changes to the shader code, mostly to get rid of JS support in shaders. JSEngine is problematic (aka crash) on other platforms than Windows. But I'm not sure when I get the time to do so.

> Another problem is the OpenGL version the program is using. You have to create two codes.

Yes, `texture2D` does not work on Intel GPU's on Windows, and `texture` does not work on NVIDIA on Windows. Don't know about AMD.

Linux is ok.

This can be fixed in the app.

if win+intel replace `texture2D` with `texture`.

**Reply by axiomgraph** [reply]
for some effects margin needs js support. removing js might not be good idea. if texture function problem get solved that would be really helpful.

**Reply by rodlie** [reply]
> for some effects margin needs js support. removing js might not be good idea

Something must be done, adjusting shader parameters will crash on Linux. Still trying to figure out why it works on Windows and not on Linux (is it Qt? is it GCC/Clang? or Friction?).

> if texture function problem get solved that would be really helpful.

I just need to find out what AMD supports on Windows (anyone with Windows+AMD GPU please test and tell me if they support `texture2D` on OpenGL 3.3).

Then I add a GPU and OS check before shader(s) code is read and replace `texture2D` if needed.

**Comment by AlexKiryanov** [comment]
Hello everyone again! Also, is it possible to add the ability to make effects work in the local coordinates of an object?
Suppose, when we rotate a texture (or object) on which an effect is added, the effect would rotate along with the texture and not remain in the global coordinates of the working field.?

I understand that I can take data from glValues and calculate it all. But of course it would be more convenient if you could choose a direct coordinate system
![pic](https://github.com/friction2d/friction/assets/65244629/fe1b0dd8-6c91-424e-99b2-68b685f7ff65)

**Reply by rodlie** [reply]
This is an issue I've noticed and will take a look at when I get the time.

**Comment by rodlie** [comment]
Btw, I have a solution for `texture2D` vs. `texture` (and motionblur fix for Intel GPU's): https://github.com/friction2d/friction/pull/86

**Reply by axiomgraph** [reply]
nice job. I loved the new keyframe button. and the problem with motion blur is that it looks like trails. If I am correct, skia is used for creating motion blur. Another problem that I have not solved yet is, for some reason, stars.frag on Windows crashes(after shadereffects initialization) friction on startup.
So that is the reason why I have not included stars.frag in version 2.0 of Enve plugins. An on-arch Linux friction drop-down menu will not work (it flickers) on Wayland. works fine with Xorg. I heared that qt-programs always had problems with Wayland. does wayland is the reason behind https://github.com/friction2d/friction/issues/64

**Reply by rodlie** [reply]
> problem with motion blur is that it looks like trails. If I am correct, skia is used for creating motion blur.

Suggestions for GPU/CPU code for motion blur is welcome. The current code for motionblur is https://github.com/friction2d/friction/blob/main/src/core/RasterEffects/motionblureffect.cpp#L269

> stars.frag on Windows crashes(after shadereffects initialization) friction on startup.

Stars from Version 1.0 seems to work on Windows here (laptop with NVIDIA), will test more computers when I get home.

> An on-arch Linux friction drop-down menu will not work (it flickers) on Wayland. works fine with Xorg. I heared that qt-programs always had problems with Wayland. does wayland is the reason behind

I don't use Wayland so don't know.

Qt apps that run native Wayland should work fine. Note that skia uses GLX, so Friction will not support (run native) Wayland due to this. The only "fix" for this is to move everything to Vulkan(?).

**Reply by axiomgraph** [reply]
I think Xorg is better. no need to move to vulkan. I had faced a lot of problems with Wayland. ShotCut Video Editor slows sometimes. obs-studio got audio cutting issues. When I switched to Xorg, most of my problems were solved.
